network/discovery/src/file.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use diem_config::{config::PeerSet, network_id::NetworkContext};
+use diem_logger::prelude::*;
+use diem_time_service::{Sleep, TimeService, TimeServiceTrait};
+use futures::{Future, Stream};
+use network::logging::NetworkSchema;
+use std::{
+    path::{Path, PathBuf},
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+pub struct FileStream {
+    network_context: Arc<NetworkContext>,
+    file_path: PathBuf,
+    time_service: TimeService,
+    interval_duration: Duration,
+    delay: Option<Pin<Box<Sleep>>>,
+}
+
+impl FileStream {
+    pub(crate) fn new(
+        network_context: Arc<NetworkContext>,
+        file_path: &Path,
+        interval_duration: Duration,
+        time_service: TimeService,
+    ) -> Self {
+        FileStream {
+            network_context,
+            file_path: file_path.to_path_buf(),
+            time_service,
+            interval_duration,
+            delay: None,
+        }
+    }
+}
+
+impl Stream for FileStream {
+    type Item = PeerSet;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Wait for delay, or add the delay for next call
+        if let Some(ref mut delay) = self.delay {
+            futures::ready!(Pin::new(delay).poll(cx));
+        }
+        self.delay = Some(Box::pin(self.time_service.sleep(self.interval_duration)));
+
+        match load_file(self.file_path.as_path()) {
+            Ok(peers) => Poll::Ready(Some(peers)),
+            Err(error) => {
+                error!(
+                    NetworkSchema::new(&self.network_context),
+                    "{} Failed to load file: {:?}", self.network_context, error
+                );
+                Poll::Ready(None)
+            }
```

Comment:
Specifically here.

Instead you may want to instead sleep for some period of time and try again later, perhaps on the next interval?

Reply:
Also might be worth adding a test for this?

Reply:
Per the [Stream](https://docs.rs/futures/0.3.15/futures/stream/trait.Stream.html) docs:
> The trait is modeled after Future, but allows poll_next to be called even after a value has been produced, yielding None once the stream has been fully exhausted.

We probably want to change this here to not return None.

Reply:
Fixed